feat: electra fork

g11tech opened the pull request:
superseeding previous rebase : 
- https://github.com/ChainSafe/lodestar/pull/6927

superceeds : 
 - https://github.com/ChainSafe/lodestar/pull/6352

g11tech committed:
feat: placeholder PR for electra

add types stub and epoch config

fix types

ensi321 committed:
feat: implement EIP-6110 (#6042)

* Add immutable in the dependencies

* Initial change to pubkeyCache

* Added todos

* Moved unfinalized cache to epochCache

* Move populating finalized cache to afterProcessEpoch

* Specify unfinalized cache during state cloning

* Move from unfinalized to finalized cache in afterProcessEpoch

* Confused myself

* Clean up

* Change logic

* Fix cloning issue

* Clean up redundant code

* Add CarryoverData in epochCtx.createFromState

* Fix typo

* Update usage of pubkeyCache

* Update pubkeyCache usage

* Fix lint

* Fix lint

* Add 6110 to ChainConfig

* Add 6110 to BeaconPreset

* Define 6110 fork and container

* Add V6110 api to execution engine

* Update test

* Add depositReceiptsRoot to process_execution_payload

* State transitioning to EIP6110

* State transitioning to EIP6110

* Light client change in EIP-6110

* Update tests

* produceBlock

* Refactor processDeposit to match the spec

* Implement processDepositReceipt

* Implement 6110 fork guard for pubkeyCache

* Handle changes in eth1 deposit

* Update eth1 deposit test

* Fix typo

* Lint

* Remove embarassing comments

* Address comments

* Modify applyDeposit signature

* Update packages/state-transition/src/cache/epochCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Update packages/state-transition/src/cache/epochCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Update packages/state-transition/src/cache/pubkeyCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Remove old code

* Rename fields in epochCache and immutableData

* Remove CarryoverData

* Move isAfter6110 from var to method

* Fix cyclic import

* Fix operations spec runner

* Fix for spec test

* Fix spec test

* state.depositReceiptsStartIndex to BigInt

* getDeposit requires cached state

* default depositReceiptsStartIndex value in genesis

* Fix pubkeyCache bug

* newUnfinalizedPubkeyIndexMap in createCachedBeaconState

* Lint

* Pass epochCache instead of pubkey2IndexFn in apis

* Address comments

* Add unit test on pubkey cache cloning

* Add unfinalizedPubkeyCacheSize to metrics

* Add unfinalizedPubkeyCacheSize to metrics

* Clean up code

* Add besu to el-interop

* Add 6110 genesis file

* Template for sim test

* Add unit test for getEth1DepositCount

* Update sim test

* Update besudocker

* Finish beacon api calls in sim test

* Update epochCache.createFromState()

* Fix bug unfinalized validators are not finalized

* Add sim test to run a few blocks

* Lint

* Merge branch 'unstable' into 611

* Add more check to sim test

* Update besu docker image instruction

* Update sim test with correct tx

* Address comment + cleanup

* Clean up code

* Properly handle promise rejection

* Lint

* Update packages/beacon-node/src/execution/engine/types.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Update comments

* Accept type undefined in ExecutionPayloadBodyRpc

* Update comment and semantic

* Remove if statement when adding finalized validator

* Comment on repeated insert on finalized cache

* rename createFromState

* Add comment on getPubkey()

* Stash change to reduce diffs

* Stash change to reduce diffs

* Lint

* addFinalizedPubkey on finalized checkpoint

* Update comment

* Use OrderedMap for unfinalized cache

* Pull out logic of deleting pubkeys for batch op

* Add updateUnfinalizedPubkeys in regen

* Update updateUnfinalizedPubkeys logic

* Add comment

* Add metrics for state context caches

* Address comment

* Address comment

* Deprecate eth1Data polling when condition is reached

* Fix conflicts

* Fix sim test

* Lint

* Fix type

* Fix test

* Fix test

* Lint

* Update packages/light-client/src/spec/utils.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Fix spec test

* Address comments

* Improve cache logic on checkpoint finalized

* Update sim test according to new cache logic

* Update comment

* Lint

* Finalized pubkey cache only update once per checkpoint

* Add perf test for updateUnfinalizedPubkeys

* Add perf test for updateUnfinalizedPubkeys

* Tweak params for perf test

* Freeze besu docker image version for 6110

* Add benchmark result

* Use Map instead of OrderedMap. Update benchmark

* Minor optimization

* Minor optimization

* Add memory test for immutable.js

* Update test

* Reduce code duplication

* Lint

* Remove try/catch in updateUnfinalizedPubkeys

* Introduce EpochCache metric

* Add historicalValidatorLengths

* Polish code

* Migrate state-transition unit tests to vitest

* Fix calculation of pivot index

* `historicalValidatorLengths` only activate post 6110

* Update sim test

* Lint

* Update packages/state-transition/src/cache/epochCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Improve readability on historicalValidatorLengths

* Update types

* Fix calculation

* Add eth1data poll todo

* Add epochCache.getValidatorCountAtEpoch

* Add todo

* Add getStateIterator for state cache

* Partial commit

* Update perf test

* updateUnfinalizedPubkeys directly modify states from regen

* Update sim test. Lint

* Add todo

* some improvements and a fix for effectiveBalanceIncrements fork safeness

* rename eip6110 to elctra

* fix electra-interop.test.ts

---------

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>
Co-authored-by: gajinder <develop@g11tech.io>

lint and tsc

small cleanup

fix rebase issue

ensi321 committed:
chore: fix CI failure due to recent merge from `unstable` (#6646)

g11tech committed:
feat: implement execution layer exits eip 7002 (#6651)

* feat: implement execution layer exits eip 7002

* lint and tsc fix

* apply feedback

* improve comment

ensi321 committed:
chore: update spec test version for electra fork (#6717)

* Update spec-test version

* Skip electra

ensi321 committed:
feat: add presets and ssz types for EIP-7549 (#6715)

* Add types

* Update unit test

* lint

* Address comments

* Address comments

* Lint

* Update packages/beacon-node/src/util/sszBytes.ts

Co-authored-by: tuyennhv <tuyen@chainsafe.io>

* Add isElectraAttestation

* Update unit test

* Update unit test

* chore: add comments for sszBytes.ts

---------

Co-authored-by: tuyennhv <tuyen@chainsafe.io>
Co-authored-by: Tuyen Nguyen <vutuyen2636@gmail.com>
Co-authored-by: Gajinder <develop@g11tech.io>

g11tech committed:
chore: fix the rebase build (#6735)

* chore: fix the rebase build

* fix test

g11tech committed:
feat: upgrade 7002 exits to withdrawal request (#6736)

* feat: upgrade 7002 exits to withdrawal request

* fix types

* fix types and references

* further fix the types references and get build passing

* update the process ops fn but needs to be extended by maxeb

ensi321 committed:
feat: implement maxEB EIP-7251 (#6539)

* feat: implement EIP-6110 (#6042)

* Add immutable in the dependencies

* Initial change to pubkeyCache

* Added todos

* Moved unfinalized cache to epochCache

* Move populating finalized cache to afterProcessEpoch

* Specify unfinalized cache during state cloning

* Move from unfinalized to finalized cache in afterProcessEpoch

* Confused myself

* Clean up

* Change logic

* Fix cloning issue

* Clean up redundant code

* Add CarryoverData in epochCtx.createFromState

* Fix typo

* Update usage of pubkeyCache

* Update pubkeyCache usage

* Fix lint

* Fix lint

* Add 6110 to ChainConfig

* Add 6110 to BeaconPreset

* Define 6110 fork and container

* Add V6110 api to execution engine

* Update test

* Add depositReceiptsRoot to process_execution_payload

* State transitioning to EIP6110

* State transitioning to EIP6110

* Light client change in EIP-6110

* Update tests

* produceBlock

* Refactor processDeposit to match the spec

* Implement processDepositReceipt

* Implement 6110 fork guard for pubkeyCache

* Handle changes in eth1 deposit

* Update eth1 deposit test

* Fix typo

* Lint

* Remove embarassing comments

* Address comments

* Modify applyDeposit signature

* Update packages/state-transition/src/cache/epochCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Update packages/state-transition/src/cache/epochCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Update packages/state-transition/src/cache/pubkeyCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Remove old code

* Rename fields in epochCache and immutableData

* Remove CarryoverData

* Move isAfter6110 from var to method

* Fix cyclic import

* Fix operations spec runner

* Fix for spec test

* Fix spec test

* state.depositReceiptsStartIndex to BigInt

* getDeposit requires cached state

* default depositReceiptsStartIndex value in genesis

* Fix pubkeyCache bug

* newUnfinalizedPubkeyIndexMap in createCachedBeaconState

* Lint

* Pass epochCache instead of pubkey2IndexFn in apis

* Address comments

* Add unit test on pubkey cache cloning

* Add unfinalizedPubkeyCacheSize to metrics

* Add unfinalizedPubkeyCacheSize to metrics

* Clean up code

* Add besu to el-interop

* Add 6110 genesis file

* Template for sim test

* Add unit test for getEth1DepositCount

* Update sim test

* Update besudocker

* Finish beacon api calls in sim test

* Update epochCache.createFromState()

* Fix bug unfinalized validators are not finalized

* Add sim test to run a few blocks

* Lint

* Merge branch 'unstable' into 611

* Add more check to sim test

* Update besu docker image instruction

* Update sim test with correct tx

* Address comment + cleanup

* Clean up code

* Properly handle promise rejection

* Lint

* Update packages/beacon-node/src/execution/engine/types.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Update comments

* Accept type undefined in ExecutionPayloadBodyRpc

* Update comment and semantic

* Remove if statement when adding finalized validator

* Comment on repeated insert on finalized cache

* rename createFromState

* Add comment on getPubkey()

* Stash change to reduce diffs

* Stash change to reduce diffs

* Lint

* addFinalizedPubkey on finalized checkpoint

* Update comment

* Use OrderedMap for unfinalized cache

* Pull out logic of deleting pubkeys for batch op

* Add updateUnfinalizedPubkeys in regen

* Update updateUnfinalizedPubkeys logic

* Add comment

* Add metrics for state context caches

* Address comment

* Address comment

* Deprecate eth1Data polling when condition is reached

* Fix conflicts

* Fix sim test

* Lint

* Fix type

* Fix test

* Fix test

* Lint

* Update packages/light-client/src/spec/utils.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Fix spec test

* Address comments

* Improve cache logic on checkpoint finalized

* Update sim test according to new cache logic

* Update comment

* Lint

* Finalized pubkey cache only update once per checkpoint

* Add perf test for updateUnfinalizedPubkeys

* Add perf test for updateUnfinalizedPubkeys

* Tweak params for perf test

* Freeze besu docker image version for 6110

* Add benchmark result

* Use Map instead of OrderedMap. Update benchmark

* Minor optimization

* Minor optimization

* Add memory test for immutable.js

* Update test

* Reduce code duplication

* Lint

* Remove try/catch in updateUnfinalizedPubkeys

* Introduce EpochCache metric

* Add historicalValidatorLengths

* Polish code

* Migrate state-transition unit tests to vitest

* Fix calculation of pivot index

* `historicalValidatorLengths` only activate post 6110

* Update sim test

* Lint

* Update packages/state-transition/src/cache/epochCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Improve readability on historicalValidatorLengths

* Update types

* Fix calculation

* Add eth1data poll todo

* Add epochCache.getValidatorCountAtEpoch

* Add todo

* Add getStateIterator for state cache

* Partial commit

* Update perf test

* updateUnfinalizedPubkeys directly modify states from regen

* Update sim test. Lint

* Add todo

* some improvements and a fix for effectiveBalanceIncrements fork safeness

* rename eip6110 to elctra

* fix electra-interop.test.ts

---------

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>
Co-authored-by: gajinder <develop@g11tech.io>

lint and tsc

small cleanup

fix rebase issue

* feat: implement EIP-6110 (#6042)

* Add immutable in the dependencies

* Initial change to pubkeyCache

* Added todos

* Moved unfinalized cache to epochCache

* Move populating finalized cache to afterProcessEpoch

* Specify unfinalized cache during state cloning

* Move from unfinalized to finalized cache in afterProcessEpoch

* Confused myself

* Clean up

* Change logic

* Fix cloning issue

* Clean up redundant code

* Add CarryoverData in epochCtx.createFromState

* Fix typo

* Update usage of pubkeyCache

* Update pubkeyCache usage

* Fix lint

* Fix lint

* Add 6110 to ChainConfig

* Add 6110 to BeaconPreset

* Define 6110 fork and container

* Add V6110 api to execution engine

* Update test

* Add depositReceiptsRoot to process_execution_payload

* State transitioning to EIP6110

* State transitioning to EIP6110

* Light client change in EIP-6110

* Update tests

* produceBlock

* Refactor processDeposit to match the spec

* Implement processDepositReceipt

* Implement 6110 fork guard for pubkeyCache

* Handle changes in eth1 deposit

* Update eth1 deposit test

* Fix typo

* Lint

* Remove embarassing comments

* Address comments

* Modify applyDeposit signature

* Update packages/state-transition/src/cache/epochCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Update packages/state-transition/src/cache/epochCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Update packages/state-transition/src/cache/pubkeyCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Remove old code

* Rename fields in epochCache and immutableData

* Remove CarryoverData

* Move isAfter6110 from var to method

* Fix cyclic import

* Fix operations spec runner

* Fix for spec test

* Fix spec test

* state.depositReceiptsStartIndex to BigInt

* getDeposit requires cached state

* default depositReceiptsStartIndex value in genesis

* Fix pubkeyCache bug

* newUnfinalizedPubkeyIndexMap in createCachedBeaconState

* Lint

* Pass epochCache instead of pubkey2IndexFn in apis

* Address comments

* Add unit test on pubkey cache cloning

* Add unfinalizedPubkeyCacheSize to metrics

* Add unfinalizedPubkeyCacheSize to metrics

* Clean up code

* Add besu to el-interop

* Add 6110 genesis file

* Template for sim test

* Add unit test for getEth1DepositCount

* Update sim test

* Update besudocker

* Finish beacon api calls in sim test

* Update epochCache.createFromState()

* Fix bug unfinalized validators are not finalized

* Add sim test to run a few blocks

* Lint

* Merge branch 'unstable' into 611

* Add more check to sim test

* Update besu docker image instruction

* Update sim test with correct tx

* Address comment + cleanup

* Clean up code

* Properly handle promise rejection

* Lint

* Update packages/beacon-node/src/execution/engine/types.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Update comments

* Accept type undefined in ExecutionPayloadBodyRpc

* Update comment and semantic

* Remove if statement when adding finalized validator

* Comment on repeated insert on finalized cache

* rename createFromState

* Add comment on getPubkey()

* Stash change to reduce diffs

* Stash change to reduce diffs

* Lint

* addFinalizedPubkey on finalized checkpoint

* Update comment

* Use OrderedMap for unfinalized cache

* Pull out logic of deleting pubkeys for batch op

* Add updateUnfinalizedPubkeys in regen

* Update updateUnfinalizedPubkeys logic

* Add comment

* Add metrics for state context caches

* Address comment

* Address comment

* Deprecate eth1Data polling when condition is reached

* Fix conflicts

* Fix sim test

* Lint

* Fix type

* Fix test

* Fix test

* Lint

* Update packages/light-client/src/spec/utils.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Fix spec test

* Address comments

* Improve cache logic on checkpoint finalized

* Update sim test according to new cache logic

* Update comment

* Lint

* Finalized pubkey cache only update once per checkpoint

* Add perf test for updateUnfinalizedPubkeys

* Add perf test for updateUnfinalizedPubkeys

* Tweak params for perf test

* Freeze besu docker image version for 6110

* Add benchmark result

* Use Map instead of OrderedMap. Update benchmark

* Minor optimization

* Minor optimization

* Add memory test for immutable.js

* Update test

* Reduce code duplication

* Lint

* Remove try/catch in updateUnfinalizedPubkeys

* Introduce EpochCache metric

* Add historicalValidatorLengths

* Polish code

* Migrate state-transition unit tests to vitest

* Fix calculation of pivot index

* `historicalValidatorLengths` only activate post 6110

* Update sim test

* Lint

* Update packages/state-transition/src/cache/epochCache.ts

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

* Improve readability on historicalValidatorLengths

* Update types

* Fix calculation

* Add eth1data poll todo

* Add epochCache.getValidatorCountAtEpoch

* Add todo

* Add getStateIterator for state cache

* Partial commit

* Update perf test

* updateUnfinalizedPubkeys directly modify states from regen

* Update sim test. Lint

* Add todo

* some improvements and a fix for effectiveBalanceIncrements fork safeness

* rename eip6110 to elctra

* fix electra-interop.test.ts

---------

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>
Co-authored-by: gajinder <develop@g11tech.io>

lint and tsc

small cleanup

* Add presets

* Update config

* Add necessary containers

* Update presets

* Update config

* Add todo comments

* Update constants and params

* Impl new process withdrawal

* Add withdrawaRequests to payload

* Add processConsolidation

* Add process withdraw request

* Update deposit and withdrawal flow

* epoch processing

* Implement churn limits

* Lint

* lint

* Update state-transition utils

* processExecutionLayerWithdrawRequest

* processConsolidation

* queueExcessActiveBalance

* isValidDepositSignature

* Add jsdoc and timer for new processEpoch functions

* Lint

* Update maxEB

* update voluntary exit

* Fix config

* Update initiateValidatorExit

* Remove churn limit in processRegistryUpdates

* Fix conflict

* Add MAX_WITHDRAWAL_REQUESTS_PER_PAYLOAD

* Reflect latest spec changes

* rebase fixes, fixes, improvements and cleanup

lint

* Upgrade ssz version

* Use sliceFrom()

* cleanup as per specs feedback

subarry

* simplify

* fix withdrawals

* remove slice

* fix the slashing quotient determination in slashvalidator

---------

Co-authored-by: harkamal <gajinder@g11.in>

ensi321 committed:
feat: beacon node process electra attestations EIP-7549  (#6738)

* Process attestations in block

* Fix check-types

* Address comments

g11tech committed:
feat: handle the EL payload sending data in deposit requests instead of deposit receipts (#6746)

ensi321 committed:
feat: implement EIP-7549 (#6689)

* initial commit

* lint

* Add getAttestingIndices and update getIndexedAttestation

* Update gossip validation

* Update attestation gossip validation

* aggregateAndProof validation

* clean up

* Validator

* Misc

* Fix the build erros

* feat: get attestations for electra block (#6732)

* feat: getAttestationsForBlock() for electra

* chore: fix lint

* fix: MAX_ATTESTATIONS_PER_GROUP_ELECTRA and address PR comments

* chore: unit test aggregateConsolidation

* Fix rebase mistake

* Address my own comment :)

---------

Co-authored-by: Navie Chan <naviechan@gmail.com>

* Fix check-types

* Address comments

---------

Co-authored-by: Nazar Hussain <nazarhussain@gmail.com>
Co-authored-by: tuyennhv <tuyen@chainsafe.io>
Co-authored-by: gajinder <develop@g11tech.io>

twoeths committed:
fix: attestation pool for electra (#6744)

* feat: attestationPool to group by slot by data root by committee index for electra

* fix: gossip validation and assert.notNull() util

* fix: remove light-client stats.html

* fix: lint and check-types

g11tech committed:
feat: update engineapi endpoints to v4 (#6747)

ensi321 committed:
feat: rename deposit receipt to deposit request for Pectra (#6748)

* Rename receipt to request

* Remove stats.html

nazarhussain committed:
test: enable spec tests related to eip-7549 (#6741)

* initial commit

* Update gossip validation

* Update attestation gossip validation

* aggregateAndProof validation

* Extend spec runner to be more flexible

* Add missing state attributes for electra

* Fix ss data types for electra spec

* Make the spec runner more flexible

* Fix the bug in process attestation

* Update the sepc test version

* clean up

* Misc

* Fix the build erros

* feat: get attestations for electra block (#6732)

* feat: getAttestationsForBlock() for electra

* chore: fix lint

* fix: MAX_ATTESTATIONS_PER_GROUP_ELECTRA and address PR comments

* chore: unit test aggregateConsolidation

* Fix rebase mistake

* Address my own comment :)

---------

Co-authored-by: Navie Chan <naviechan@gmail.com>

* Fix check-types

* Address comments

* Fix the build erros

* Extend spec runner to be more flexible

* Add missing state attributes for electra

* Fix ss data types for electra spec

* Make the spec runner more flexible

* Fix the bug in process attestation

* Update the sepc test version

* Fix rebase issue

* Update committee index count check

---------

Co-authored-by: NC <adrninistrator1@protonmail.com>
Co-authored-by: Navie Chan <naviechan@gmail.com>
Co-authored-by: tuyennhv <tuyen@chainsafe.io>

ensi321 committed:
fix: fix e2e test in electra-fork (#6751)

Update spec version

g11tech committed:
feat: get the basic integration working with the ethereumjs electra build (#6752)

g11tech committed:
feat: apply some fixes and hacks to get the single node devnet working with fork transition (#6754)

twoeths committed:
fix: get aggregate and proofs signature sets (#6757)

fix: get signature for SignedAggregateAndProof based on fork

nazarhussain committed:
test(spec): fix attestors slashing specs for electra fork (#6758)

* Fix attester slashing specs for electra

* Remove unused import

* Add code comment

* Update the expression

* Update the fork check

g11tech committed:
chore: fix types and lint (#6750)

* chore: fix types and lint

* fx

* type and lint fix

ensi321 committed:
fix: fix electra genesis spec test (#6764)

* process pending deposit from eth1

* Fix the genesis params

* fix

* Fix

* clean up

---------

Co-authored-by: Nazar Hussain <nazarhussain@gmail.com>

ensi321 committed:
feat: support missing electra spec test (#6765)

Add spec test

nflaig committed:
test: fix ssz types in fork_choice spec tests (#6767)

jeluard committed:
chore: update EffectiveBalanceIncrements type (#6763)

* chore: update EffectiveBalanceIncrements type

* chore: remove now irrelevant tests

ensi321 committed:
Fix ssz_static

matthewkeil committed:
fix: inline sourcemaps to help with debugging (#6768)

nazarhussain committed:
fix: additional epoch calculation logic for consolidation churn (#6770)

Fix the chunk limit logic

twoeths committed:
fix: electra fork transition spec tests (#6769)

* fix: electra fork transition

* fix: merge issue

* chore: remove unwanted change

nflaig committed:
test: fix ssz_static spec tests for all forks (#6771)

nazarhussain committed:
chore(spec): remove the skip specs for electra (#6772)

Remove the skip spec for electra

ensi321 committed:
fix: use mutable validator object (#6774)

Use mutable validator object

matthewkeil committed:
test: fix balance spec tests (#6777)

* fix: remove epochCache.balances and invalid MAX_EFFECTIVE_BALANCE check

* fix: update rewardsAndPenalties balance updates

* docs: add comment to check epochTransitionCache

ensi321 committed:
fix: effective balance cache is not in sync with validator effective balance (#6780)

Update eb cache at fork transition

ensi321 committed:
fix: make electra-fork passes lint and check-types (#6785)

fix lint and check-types

nflaig committed:
fix: update data format of WithdrawalRequestV1 (#6789)

ensi321 committed:
fix: publish attestations with non-zero committee index (#6790)

Fix publishing att with non-zero comm index

ensi321 committed:
fix: validator monitor summaries should not render during epoch 0 (#6791)

Skip render summary at epoch 0

ensi321 committed:
fix: attestation duty validation (#6792)

* fix attestation duty validation

* Update packages/validator/src/services/validatorStore.ts

Co-authored-by: twoeths <tuyen@chainsafe.io>

* Update packages/validator/src/services/validatorStore.ts

---------

Co-authored-by: twoeths <tuyen@chainsafe.io>
Co-authored-by: Cayman <caymannava@gmail.com>

wemeetagain committed:
fix: align BeaconBlockBody and BlindedBeaconBlockBody (#6782)

* fix: align BeaconBlockBody and BlindedBeaconBlockBody

* Remove type hacks in test

---------

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

jeluard committed:
test: improve ssz tests consistency (#6776)

* test: improve ssz tests consistency

* chore: address comments

twoeths committed:
fix: batch validation for electra attestations (#6788)

nflaig committed:
fix: update withdrawal request container to match consensus spec (#6797)

twoeths committed:
fix: get seen AttData key from SignedAggregateAndProof electra (#6802)

* fix: get seen AttData key from SignedAggregateAndProof electra

* chore: revert the naming change to COMMITTEE_BITS_SIZE and add comment

* fix: add toBase64() util

nflaig committed:
test: only skip ssz_static tests associated to missing type (#6798)

* test: only skip ssz_static tests associated to missing type

* More detailed error message if type is not defined

g11tech committed:
chore: types and lint fixes (#6819)

ensi321 committed:
feat: add engine_getPayloadBodiesByHash and ByRange V2 (#6852)

* Add ByHash and ByRange V2

* Fix build issue

* Fix CI error

nflaig committed:
fix: align WithdrawalRequestV1 field names with latest spec (#6877)

ensi321 committed:
feat: move attestation committee at the end of attestation (#6883)

g11tech committed:
rebase fixes

chore: fix bls and blst versioning

fix: add ForkName.electra to ForkBlobsInfo

some api header lookup fixes

more api fixes

make the api data safe

Co-authored-by: matthewkeil <me@matthewkeil.com>

ensi321 committed:
feat: handle exited/exiting validators during top up (#6880)

* Handle exiting validataor

* lint

* Add todo

ensi321 committed:
feat: add EL triggered consolidation and remove `ExecutionLayer` prefix (#6865)

* init commit

* Add consolidation request

* fix lint

* Remove ExecutionLayer prefix

* Address comment

* Fix verification logic

* lint

* Add todo

ensi321 committed:
feat: support electra devnet-1 (#6892)

* Update spec test version

* Use new max effective balance

* Relax loop breaking condition for `computeProposerIndex`

* fix remaining

* check-types & lint

* Skip invalid test

* Fix rebase + lint

* Remove early return statement in `computeProposers`

* Address comment

* Address comment

g11tech committed:
fix: electra rebase fixes on the new generic typing model

lint & type fix

lint

g11tech committed:
feat: add and parse consolidation requests from engine api

get the signing log

g11tech committed:
feat: make produce block/signed block contents api multifork

g11tech committed:
fix: electra rebase fixes on the allforks type refactor

tests fixes

g11tech committed:
chore: rebase fixes

lint

lint and type fix

update spec version

fix tests

fix spec test

Update packages/beacon-node/src/chain/stateCache/types.ts

Co-authored-by: Cayman <caymannava@gmail.com>

Update packages/beacon-node/src/chain/stateCache/types.ts

Co-authored-by: Cayman <caymannava@gmail.com>

Update packages/beacon-node/src/chain/blocks/types.ts

Co-authored-by: Cayman <caymannava@gmail.com>

Update packages/types/src/utils/typeguards.ts

Co-authored-by: Cayman <caymannava@gmail.com>

Variable naming

fix the ForkElectra

fix build

fix spec test

Co-authored-by: Cayman <caymannava@gmail.com>

wemeetagain commented on packages/api/src/beacon/routes/beacon/block.ts:
```diff
@@ -297,11 +290,12 @@ export function getDefinitions(config: ChainForkConfig): RouteDefinitions<Endpoi
           const slot = isSignedBlockContents(signedBlockOrContents)
             ? signedBlockOrContents.signedBlock.message.slot
             : signedBlockOrContents.message.slot;
+          const fork = config.getForkName(slot);
```
looks like the changes here in this file are all refactors? could be PR'd into unstable.

wemeetagain commented on packages/api/src/beacon/routes/beacon/pool.ts:
```diff
@@ -183,8 +191,10 @@ export function getDefinitions(_config: ChainForkConfig): RouteDefinitions<Endpo
       resp: {
-        data: AttestationListType,
-        meta: EmptyMetaCodec,
+        data: WithVersion((fork) =>
```
will re-review this file after https://github.com/ChainSafe/lodestar/pull/6951

wemeetagain commented on packages/api/src/beacon/routes/validator.ts:
```diff
@@ -801,33 +800,92 @@ export function getDefinitions(_config: ChainForkConfig): RouteDefinitions<Endpo
       req: {
-        writeReq: ({attestationDataRoot, slot}) => ({
-          query: {attestation_data_root: toHexString(attestationDataRoot), slot},
+        writeReq: ({attestationDataRoot, slot, committeeIndex}) => ({
```
needs to only add a v2 here and below, not change v1

wemeetagain commented on packages/api/test/unit/beacon/oapiSpec.test.ts:
```diff
@@ -86,6 +86,11 @@ const ignoredTopics = [
    topic block_gossip not implemented
    */
   "block_gossip",
+
```
may be worth pulling a specific commit of the dev branch of beacon-apis so we can actually test (I'm assuming beacon-apis will not cut an actual release until X months from now)

nflaig replied:
dev branch does not work as we are generally using the json file of the beacon spec which is only generated on release via swagger-cli. I have created a repo to host the json spec for now, this approach is already used in the open v2 apis branch, see https://github.com/ChainSafe/lodestar/pull/6951/files#diff-e95494735125652b2939787df6466e32cd940c1fbd0074200f616d87c6d6e6f6R26.

I have also asked in the ETH R&D api channel if it's worth to cut a pre-release (similar to CL spec) but it does not seem like there is much interest as we are the only client that has api spec tests.

wemeetagain replied:
Definitely in favor of a pre-release cycle that tracks the consensus-specs pre-release cycle

Maybe the reason we're the only client that has API spec tests is that the beacon API isn't very useful for testing since it rarely has releases that match the consensus-specs :joy:

nflaig replied:
I wanted to look into beacon api tooling soon because right now it's really outdated, and keymanager api can't even be properly released right now since the release process is like 2 years old and not on par with how it's done on the beacon api spec.

I will try to push pre-releases as we definitely benefit from this, not only for new apis added for a hard fork but in general.

nflaig replied:
We might have pre-releases soon (see https://github.com/ethereum/beacon-APIs/pull/463)

wemeetagain commented on packages/beacon-node/src/chain/chain.ts:
```diff
@@ -1098,6 +1101,40 @@ export class BeaconChain implements IBeaconChain {
+    const cpEpoch = cp.epoch;
+    const electraEpoch = headState?.config.ELECTRA_FORK_EPOCH ?? Infinity;
+
+    if (headState === null) {
```
consider refactoring here. the head state should always be available here.

wemeetagain commented on packages/beacon-node/src/chain/genesis/genesis.ts:
```diff
@@ -174,7 +174,13 @@ export class GenesisBuilder implements IGenesisBuilder {
       });
 
-    const {activatedValidatorCount} = applyDeposits(this.config, this.state, newDeposits, this.depositTree);
+    const {activatedValidatorCount} = applyDeposits(
```
does this function signature need to be updated?

first line of function can be `const forkSeq = config.getForkSeq(state.slot);`

wemeetagain commented on packages/beacon-node/src/chain/opPools/opPool.ts:
```diff
@@ -207,7 +208,8 @@ export class OpPool {
     const endAttesterSlashings = stepsMetrics?.startTimer();
-    const attesterSlashings: phase0.AttesterSlashing[] = [];
+    const attesterSlashings: AttesterSlashing[] = [];
+    const maxAttesterSlashing = stateFork >= ForkSeq.electra ? MAX_ATTESTER_SLASHINGS_ELECTRA : MAX_ATTESTER_SLASHINGS;
```
```suggestion
    const maxAttesterSlashings = stateFork >= ForkSeq.electra ? MAX_ATTESTER_SLASHINGS_ELECTRA : MAX_ATTESTER_SLASHINGS;
```

wemeetagain commented on packages/beacon-node/src/chain/chain.ts:
```diff
@@ -1098,6 +1101,40 @@ export class BeaconChain implements IBeaconChain {
     }
+
+    const cpEpoch = cp.epoch;
+    const electraEpoch = headState?.config.ELECTRA_FORK_EPOCH ?? Infinity;
```
```suggestion
    const electraEpoch = this.config.ELECTRA_FORK_EPOCH;
```

nflaig replied:
assigning a CONSTANT to a normal variable seems strange, should consider removing `electraEpoch`

wemeetagain commented on packages/beacon-node/src/chain/regen/queued.ts:
```diff
@@ -195,6 +195,54 @@ export class QueuedStateRegenerator implements IStateRegenerator {
+    const addTimer = this.metrics?.regenFnAddPubkeyTime.startTimer();
+
+    // We only need to add pubkeys to any one of the states since the finalized caches is shared globally across all states
+    const firstState = (states.next().value ?? cpStates.next().value) as CachedBeaconStateAllForks | undefined;
```
not worth refactoring now, but there should be a way to get the head state thats guaranteed

wemeetagain commented on packages/beacon-node/src/chain/stateCache/fifoBlockStateCache.ts:
```diff
@@ -190,6 +190,10 @@ export class FIFOBlockStateCache implements BlockStateCache {
   }
 
+  getStates(): IterableIterator<CachedBeaconStateAllForks> {
+    throw new Error("Method not implemented.");
```
add TODO or implement

wemeetagain commented on packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts:
```diff
@@ -594,6 +594,10 @@ export class PersistentCheckpointStateCache implements CheckpointStateCache {
   }
 
+  getStates(): IterableIterator<CachedBeaconStateAllForks> {
+    throw new Error("Method not implemented.");
```
add TODO or implement

wemeetagain commented on packages/beacon-node/src/chain/regen/queued.ts:
```diff
@@ -195,6 +195,54 @@ export class QueuedStateRegenerator implements IStateRegenerator {
+  updateUnfinalizedPubkeys(validators: UnfinalizedPubkeyIndexMap): void {
+    let numStatesUpdated = 0;
+    const states = this.blockStateCache.getStates();
+    const cpStates = this.checkpointStateCache.getStates();
```
It seems that what we really want is all EpochCache's in memory.
Not sure what the refactoring path is here, but just worth noting.

wemeetagain commented on packages/beacon-node/src/chain/stateCache/types.ts:
```diff
@@ -33,6 +33,7 @@ export interface BlockStateCache {
   prune(headStateRootHex: RootHex): void;
   deleteAllBeforeEpoch(finalizedEpoch: Epoch): void;
   dumpSummary(): routes.lodestar.StateCacheItem[];
+  getStates(): IterableIterator<CachedBeaconStateAllForks>; // Expose beacon states stored in cache. Use with caution
```
```suggestion
  /** Expose beacon states stored in cache. Use with caution */
  getStates(): IterableIterator<CachedBeaconStateAllForks>;
```

wemeetagain commented on packages/beacon-node/src/chain/stateCache/types.ts:
```diff
@@ -74,6 +75,7 @@ export interface CheckpointStateCache {
   processState(blockRootHex: RootHex, state: CachedBeaconStateAllForks): Promise<number>;
   clear(): void;
   dumpSummary(): routes.lodestar.StateCacheItem[];
+  getStates(): IterableIterator<CachedBeaconStateAllForks>; // Expose beacon states stored in cache. Use with caution
```
```suggestion
  /** Expose beacon states stored in cache. Use with caution */
  getStates(): IterableIterator<CachedBeaconStateAllForks>;
```

wemeetagain commented on packages/beacon-node/src/chain/validation/aggregateAndProof.ts:
```diff
@@ -163,11 +179,16 @@ async function validateAggregateAndProof(
   const attestingIndices = aggregate.aggregationBits.intersectValues(committeeIndices);
-  const indexedAttestation: phase0.IndexedAttestation = {
+
+  const indexedAttestationContent = {
```
```suggestion
  const indexedAttestation: IndexedAttestation = {
```

wemeetagain commented on packages/types/src/utils/typeguards.ts:
```diff
@@ -66,3 +67,9 @@ export function isSignedBlockContents<F extends ForkBlobs>(
+
+export function isElectraAttestation<F extends ForkName.electra>(
+  attestation: Attestation
+): attestation is Attestation<F> {
```
```suggestion
export function isElectraAttestation(
  attestation: Attestation
): attestation is Attestation<ForkPostElectra> {
```

wemeetagain commented on packages/state-transition/src/util/validator.ts:
```diff
@@ -47,3 +53,48 @@ export function getActivationChurnLimit(config: ChainForkConfig, fork: ForkSeq,
+  if (hasCompoundingWithdrawalCredential(withdrawalCredentials)) {
+    return MAX_EFFECTIVE_BALANCE_ELECTRA;
+  } else {
+    return MIN_ACTIVATION_BALANCE;
```
```suggestion
    return MAX_EFFECTIVE_BALANCE;
```

ensi321 replied:
`MIN_ACTIVATION_BALANCE` is correct as per spec: https://github.com/ethereum/consensus-specs/blob/b394b27e2bab42dc6132148634d68d16bca652c3/specs/electra/beacon-chain.md?plain=1#L544

wemeetagain commented on packages/state-transition/src/util/validator.ts:
```diff
@@ -47,3 +53,48 @@ export function getActivationChurnLimit(config: ChainForkConfig, fork: ForkSeq,
+export function getBalanceChurnLimit(state: CachedBeaconStateElectra): number {
+  const churnLimitByTotalActiveBalance = Math.floor(
+    (state.epochCtx.totalActiveBalanceIncrements / state.config.CHURN_LIMIT_QUOTIENT) * EFFECTIVE_BALANCE_INCREMENT
+  ); // TODO Electra: verify calculation
```
should verify and remove TODO

wemeetagain commented on packages/state-transition/src/util/validator.ts:
```diff
@@ -47,3 +53,48 @@ export function getActivationChurnLimit(config: ChainForkConfig, fork: ForkSeq,
+/**
+ * Get combined churn limit of activation-exit and consolidation
+ */
+export function getBalanceChurnLimit(state: CachedBeaconStateElectra): number {
```
consider improving this function signature to support multiple forks, same with below

eg:
`getBalanceChurnLimit(epochCtx: EpochCache): number`

wemeetagain commented on packages/state-transition/src/util/electra.ts:
```diff
@@ -0,0 +1,104 @@
+  balance: number,
+  epoch: number
+): boolean {
+  const {withdrawableEpoch, withdrawalCredentials: withdrawalCredential} = validatorCredential;
```
```suggestion
  const {withdrawableEpoch, withdrawalCredentials} = validatorCredential;
```

same below

wemeetagain commented on packages/state-transition/src/cache/epochCache.ts:
```diff
@@ -572,27 +634,78 @@ export class EpochCache {
+  }
+
+  /**
+   * Return a single Uint32Array representing concatted committees of indices
```
may be worth adding a TODO to avoid concatting here, this will get expensive if it is actually used frequently with multiple indices

twoeths replied:
this could be done in the shuffling refactor #6938 or leave TODO there
cc @matthewkeil

matthewkeil replied:
I look at adding this now

wemeetagain commented on packages/state-transition/src/cache/epochCache.ts:
```diff
@@ -692,6 +805,31 @@ export class EpochCache {
+  /**
+   * Return indices of validators who attestested in `attestation`
+   */
+  getAttestingIndices(fork: ForkSeq, attestation: Attestation): number[] {
```
can this function use `this.isAfterElectra` instead of `fork`?

same with `getIndexedAttestation`

wemeetagain commented on packages/state-transition/src/cache/epochCache.ts:
```diff
@@ -838,15 +1042,53 @@ export class EpochCache {
   }
+
+  isAfterElectra(): boolean {
+    return this.epoch >= (this.config.ELECTRA_FORK_EPOCH ?? Infinity);
```
do we need the ?? Infinity here?

wemeetagain commented on packages/state-transition/src/cache/epochCache.ts:
```diff
@@ -838,15 +1042,53 @@ export class EpochCache {
     this.effectiveBalanceIncrements[index] = Math.floor(effectiveBalance / EFFECTIVE_BALANCE_INCREMENT);
   }
+
+  isAfterElectra(): boolean {
```
consider renaming to `isPostElectra` to match `ForkPostElectra`

wemeetagain replied:
this should probably be private

ensi321 replied:
> this should probably be private

I think `isPostElectra()` provides a convenient way for eth1 deposit tracker to determine when to stop polling data
https://github.com/ChainSafe/lodestar/blob/3f4b2cb9c37f3032bcfc3d947aee5d22b3f03a7a/packages/beacon-node/src/eth1/eth1DepositDataTracker.ts#L131

wemeetagain commented on packages/state-transition/src/cache/pubkeyCache.ts:
```diff
@@ -40,7 +55,7 @@ export class PubkeyIndexMap {
   }
 
-  set(key: Uint8Array, value: ValidatorIndex): void {
+  set(key: Uint8Array | PubkeyHex, value: ValidatorIndex): void {
```
why is this changed?

wemeetagain commented on packages/state-transition/src/cache/types.ts:
```diff
@@ -8,6 +8,7 @@ export type BeaconStateAltair = CompositeViewDU<SSZTypesFor<ForkName.altair, "Be
 export type BeaconStateBellatrix = CompositeViewDU<SSZTypesFor<ForkName.bellatrix, "BeaconState">>;
 export type BeaconStateCapella = CompositeViewDU<SSZTypesFor<ForkName.capella, "BeaconState">>;
 export type BeaconStateDeneb = CompositeViewDU<SSZTypesFor<ForkName.deneb, "BeaconState">>;
+export type BeaconStateElectra = CompositeViewDU<SSZTypesFor<ForkName.electra, "BeaconState">>;
```
maybe make a TODO to clean up these types (like we've done with our basic forky ssz types)

This would involve having something like:
```ts
export type BeaconStateViewDU<F extends ForkName = ForkName> = CompositeViewDU<SSZTypesFor<F, "BeaconState">>
```

as well as
```ts
export type CachedBeaconState<F extends ForkName = ForkName> = BeaconStateViewDU<F> & BeaconStateCache;
```

wemeetagain commented on packages/api/package.json:
```diff
@@ -71,7 +71,7 @@
   "dependencies": {
     "@chainsafe/persistent-merkle-tree": "^0.7.1",
-    "@chainsafe/ssz": "^0.15.1",
+    "@chainsafe/ssz": "^0.16.0",
```
this could be pulled out into a PR to unstable

wemeetagain commented on packages/beacon-node/src/chain/blocks/types.ts:
```diff
@@ -36,7 +36,7 @@ export enum GossipedInputType {
 type BlobsCacheMap = Map<number, {blobSidecar: deneb.BlobSidecar; blobBytes: Uint8Array | null}>;
 
-type ForkBlobsInfo = {fork: ForkName.deneb};
+type ForkBlobsInfo = {fork: ForkName.deneb | ForkName.electra};
```
```suggestion
type ForkBlobsInfo = {fork: ForkBlobs};
```

wemeetagain reviewed (commented):
partway thru

wemeetagain commented on packages/beacon-node/src/metrics/metrics/lodestar.ts:
```diff
@@ -1373,6 +1389,21 @@ export function createLodestarMetrics(
+    regenFnAddPubkeyTime: register.histogram({
+      name: "lodestar_regen_fn_add_pubkey_time_seconds",
+      help: "Historgram of time spent on adding pubkeys to all state cache items in seconds",
+      buckets: [0.01, 0.1, 0.5, 1, 2, 5],
+    }),
+    regenFnDeletePubkeyTime: register.histogram({
+      name: "lodestar_regen_fn_delete_pubkey_time_seconds",
+      help: "Histrogram of time spent on deleting pubkeys from all state cache items in seconds",
+      buckets: [0.01, 0.1, 0.5, 1, 2, 5],
```
I don't think 2 or 5 seconds will be useful here (but may be worth investigating)

wemeetagain commented on packages/beacon-node/src/network/gossip/topic.ts:
```diff
@@ -128,9 +128,10 @@ export function sszDeserialize<T extends GossipTopic>(topic: T, serializedData:
-export function sszDeserializeAttestation(serializedData: Uint8Array): phase0.Attestation {
+export function sszDeserializeAttestation(fork: ForkName, serializedData: Uint8Array): Attestation {
+  const sszType = sszTypesFor(fork).Attestation;
   try {
-    return ssz.phase0.Attestation.deserialize(serializedData);
+    return sszType.deserialize(serializedData);
```
no reason not to make this a one-liner

wemeetagain commented on packages/config/src/forkConfig/types.ts:
```diff
@@ -21,11 +21,14 @@ export type ForkConfig = {
   getForkInfo(slot: Slot): ForkInfo;
-
+  /** Get the hard-fork info for the active fork at `epoch` */
+  getForkInfoFromEpoch(epoch: Epoch): ForkInfo;
   /** Get the hard-fork name at a given slot */
   getForkName(slot: Slot): ForkName;
   /** Get the hard-fork sequence number at a given slot */
   getForkSeq(slot: Slot): ForkSeq;
+  /** Get the hard-fork sequence number at a given epoch */
+  getForkSeqFromEpoch(epoch: Epoch): ForkSeq;
```
slightly more concise to name as `getFork*AtEpoch`

wemeetagain commented on packages/params/src/forkName.ts:
```diff
@@ -78,3 +80,16 @@ export const forkBlobs = exclude(forkAll, [ForkName.phase0, ForkName.altair, For
   return isForkWithdrawals(fork) && fork !== ForkName.capella;
 }
+
+export type ForkPreElectra = ForkPreBlobs | ForkName.deneb;
+export type ForkElectra = Exclude<ForkName, ForkPreElectra>;
+export const forkElectra = exclude(forkAll, [
+  ForkName.phase0,
+  ForkName.altair,
+  ForkName.bellatrix,
+  ForkName.capella,
+  ForkName.deneb,
+]);
+export function isForkElectra(fork: ForkName): fork is ForkElectra {
+  return isForkBlobs(fork) && fork !== ForkName.deneb;
+}
```
Can you name these `ForkPostElectra`, `forkPostElectra` and `isForkPostElectra`?

These types / values are meant to represent "all forks starting with electra and there afterwards", not just "the singular fork named electra".

wemeetagain commented on packages/state-transition/package.json:
```diff
@@ -61,13 +61,17 @@
     "@lodestar/utils": "^1.20.2",
-    "bigint-buffer": "^1.1.5"
+    "bigint-buffer": "^1.1.5",
+    "buffer-xor": "^2.0.2",
+    "immutable": "^4.3.2"
+  },
+  "devDependencies": {
+    "@types/buffer-xor": "^2.0.0"
```
is buffer-xor still used? I think its gone

wemeetagain commented on packages/types/src/utils/typeguards.ts:
```diff
@@ -66,3 +67,7 @@ export function isSignedBlockContents<F extends ForkBlobs>(
   return (data as SignedBlockContents<F>).kzgProofs !== undefined;
 }
+
+export function isElectraAttestation<F extends ForkElectra>(attestation: Attestation): attestation is Attestation<F> {
```
```suggestion
export function isElectraAttestation(attestation: Attestation): attestation is Attestation<ForkPostElectra> {
```

I'd be interested in seeing where we use this, we maybe should be using fork / slot to determine which type it is.

ensi321 replied:
Hmmm need more thought on determining electra attestation based on slot vs presence of committee bits. 

If we receive a conflicting attestation eg. pre-electra slot with committee bits, post-electra slot with no committee bits, we need a way to handle it properly.

wemeetagain commented on yarn.lock:
```diff
@@ -313,6 +313,11 @@
   resolved "https://registry.yarnpkg.com/@chainsafe/as-sha256/-/as-sha256-0.4.1.tgz#cfc0737e25f8c206767bdb6703e7943e5d44513e"
   integrity sha512-IqeeGwQihK6Y2EYLFofqs2eY2ep1I2MvQXHzOAI+5iQN51OZlUkrLgyAugu2x86xZewDk5xas7lNczkzFzF62w==
 
+"@chainsafe/as-sha256@^0.4.2":
```
somewhere as-sha256 0.4.1 is still being used

ensi321 committed:
feat: add Electra attestation V2 endpoints (#6951)

* Initial commit

* getAggregatedAttestationV2

* Lint

* Fix minor flaw

* Add publishAggregateAndProofsV2

* Fix spelling

* Fix CI

* Fix spec test

* Clean up events api

* Run against latest beacon api spec

* Revert changes to emitted events

* Update packages/api/src/beacon/routes/beacon/pool.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

* Update packages/api/src/beacon/routes/beacon/pool.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

* Update packages/api/src/beacon/routes/beacon/pool.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

* Update packages/api/src/beacon/routes/beacon/pool.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

* Address comment

* Add api stub back

* Add todos

* Review PR

* Fix rebase

* Lint

---------

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

ensi321 committed:
Update packages/beacon-node/src/chain/validation/aggregateAndProof.ts

Co-authored-by: Cayman <caymannava@gmail.com>

Update packages/types/src/utils/typeguards.ts

Co-authored-by: Cayman <caymannava@gmail.com>

Partial address comment

Partial address comment

Co-authored-by: Cayman <caymannava@gmail.com>

g11tech committed:
fix: fix the devnet2 bug of invalid stateroot calc post consolidation

g11tech committed:
chore: further rebase fixes to unstable

lint fixes

lint

Fix browser test

Fix unit test

---------

Co-authored-by: NC <17676176+ensi321@users.noreply.github.com>

nflaig commented on packages/beacon-node/src/execution/engine/types.ts:
```diff
@@ -118,9 +126,23 @@ type ExecutionPayloadRpcWithValue = {
+export type ExecutionPayloadBodyRpc = {
+  transactions: DATA[];
+  withdrawals: WithdrawalV1[] | null | undefined;
+  // currently there is a discepancy between EL and CL field name references for deposit requests
+  // its likely CL receipt will be renamed to requests
```
stale comment?

nflaig commented on packages/beacon-node/src/execution/engine/types.ts:
```diff
@@ -212,6 +241,14 @@ export function serializeExecutionPayload(fork: ForkName, data: ExecutionPayload
     payload.excessBlobGas = numToQuantity(excessBlobGas);
   }
 
+  // ELECTRA adds depositRequests/depositRequests to the ExecutionPayload
```
```suggestion
  // ELECTRA adds new fields to the ExecutionPayload
```

nflaig commented on packages/beacon-node/src/execution/engine/types.ts:
```diff
@@ -297,6 +334,35 @@ export function parseExecutionPayload(
   }
 
+  if (ForkSeq[fork] >= ForkSeq.electra) {
+    // electra adds depositRequests/depositRequests
```
```suggestion
    // electra adds new fields
```

nflaig commented on packages/beacon-node/src/metrics/metrics/beacon.ts:
```diff
@@ -120,6 +120,13 @@ export function createBeaconMetrics(register: RegistryMetricCreator) {
 
+    headState: {
+      unfinalizedPubkeyCacheSize: register.gauge({
+        name: "head_state_unfinalized_pubkey_cache_size",
```
prefix in metric name is missing

```suggestion
        name: "beacon_head_state_unfinalized_pubkey_cache_size",
```

I also noticed all of the metrics added in this branch are not used anywhere

nflaig commented on packages/beacon-node/src/metrics/metrics/lodestar.ts:
```diff
@@ -374,6 +374,17 @@ export function createLodestarMetrics(
 
+    epochCache: {
+      finalizedPubkeyDuplicateInsert: register.gauge({
+        name: "lodestar_epoch_cache_finalized_pubkey_duplicate_insert",
```
```suggestion
        name: "lodestar_epoch_cache_finalized_pubkey_duplicate_insert_total",
```

nflaig commented on packages/beacon-node/src/metrics/metrics/lodestar.ts:
```diff
@@ -374,6 +374,17 @@ export function createLodestarMetrics(
+        help: "Total count of duplicate insert of finalized pubkeys",
+      }),
+      newUnFinalizedPubkey: register.gauge({
+        name: "lodestar_epoch_cache_new_unfinalized_pubkey",
```
Which should try to follow [best practices](https://prometheus.io/docs/practices/naming/) for new metrics, this will also improve consistency

```suggestion
        name: "lodestar_epoch_cache_new_unfinalized_pubkey_total",
```

nflaig commented on packages/beacon-node/test/utils/state.ts:
```diff
@@ -92,6 +95,12 @@ export function generateState(
 
+  if (forkSeq >= ForkSeq.electra) {
+    const stateElectra = state as electra.BeaconState;
+    stateElectra.depositRequestsStartIndex = 2023n;
```
why is this set to 2023? strange test value

nflaig commented on packages/flare/src/cmds/selfSlashAttester.ts:
```diff
@@ -130,7 +130,7 @@ export async function selfSlashAttesterHandler(args: SelfSlashArgs): Promise<voi
     };
 
-    (await client.beacon.submitPoolAttesterSlashings({attesterSlashing})).assertOk();
+    (await client.beacon.submitPoolAttesterSlashingsV2({attesterSlashing})).assertOk();
```
I think we are good with using v2 here right away

nflaig commented on packages/params/src/presets/mainnet.ts:
```diff
@@ -118,4 +118,21 @@ export const mainnetPreset: BeaconPreset = {
+  MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP: 8,
+  // 2**11 * 10**9 (= 2,048,000,000,000) Gwei
+  MAX_EFFECTIVE_BALANCE_ELECTRA: 2048000000000,
+  // 2**16 (= 65536)
```
stray comment? (same in minimal.ts)

nflaig reviewed (commented):
halfway through reviewing the changes, opened a PR to fix and cleanup a few things https://github.com/ChainSafe/lodestar/pull/7015

twoeths committed:
fix: cached balances in epoch transition (#7018)

* fix: update cached balances in epoch transition

* fix: more comments in processEffectiveBalanceUpdates()

nflaig commented on packages/state-transition/src/block/processWithdrawalRequest.ts:
```diff
@@ -0,0 +1,99 @@
+  state: CachedBeaconStateElectra,
+  withdrawalRequest: electra.WithdrawalRequest
+): void {
+  const amount = Number(withdrawalRequest.amount);
```
as per type, this conversion should not be required
```suggestion
  const amount = withdrawalRequest.amount;
```

nflaig commented on packages/state-transition/src/util/epoch.ts:
```diff
@@ -39,6 +41,60 @@ export function computeActivationExitEpoch(epoch: Epoch): Epoch {
   return epoch + 1 + MAX_SEED_LOOKAHEAD;
 }
 
+export function computeExitEpochAndUpdateChurn(state: CachedBeaconStateElectra, exitBalance: Gwei): number {
```
might wanna move these functions to a different file as they don't really fit here. Maybe `packages/state-transition/src/util/electra.ts`?

nflaig commented on packages/state-transition/test/perf/util/shufflings.test.ts:
```diff
@@ -28,7 +28,8 @@ describe("epoch shufflings", () => {
     fn: () => {
       const epochSeed = getSeed(state, state.epochCtx.nextShuffling.epoch, DOMAIN_BEACON_PROPOSER);
-      computeProposers(epochSeed, state.epochCtx.nextShuffling, state.epochCtx.effectiveBalanceIncrements);
+      const fork = state.config.getForkSeq(state.slot);
```
might wanna add `state.fork`, this pattern is quite frequent

nflaig commented on packages/validator/test/unit/services/attestation.test.ts:
```diff
@@ -68,6 +71,7 @@ describe("AttestationService", function () {
           chainHeadTracker,
           syncingStatusTracker,
           null,
+          createChainForkConfig(config),
```
we might wanna run these tests pre-/post-electra, can be addressed in the PR that enabled v1 apis pre-electra

nflaig replied:
added post-electra tests in https://github.com/ChainSafe/lodestar/pull/7024

nflaig commented on packages/validator/test/unit/utils/interopConfigs.ts:
```diff
@@ -383,6 +402,13 @@ export const tekuHoleskyConfig = {
   DOMAIN_AGGREGATE_AND_PROOF: "0x06000000",
   CHURN_LIMIT_QUOTIENT: "65536",
   BLS_WITHDRAWAL_PREFIX: "0x00",
+  MAX_EFFECTIVE_BALANCE_ELECTRA: "2048000000000",
```
are these updated based on what Teku actually returns or did we just add this to make tests pass? If those are not actual values from another client the interop config tests are pointless

nflaig reviewed (commented):
Review of second half, opened another PR to clean up a few things https://github.com/ChainSafe/lodestar/pull/7019

nflaig committed:
chore: review electra branch - part 1 (#7015)

nflaig committed:
chore: review electra branch - part 2 (#7019)

nflaig committed:
feat: support post-electra attestation and attester slashing events (#7026)

ensi321 committed:
feat: pre-electra support from attestation pool (#6998)

* Initial commit

* Update packages/beacon-node/src/chain/chain.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

* Update packages/beacon-node/src/api/impl/validator/index.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

* Update packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

* address comment

* Add unit test for attestation pool

* fix: getSeenAttDataKey apis (#7009)

* fix: getSeenAttDataKey apis

* chore: use ForkName instead of ForkSeq

* Update packages/beacon-node/src/util/sszBytes.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

* Update packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

* address comment

* Update error message

* Update packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

* address comment

* Move determining post-electra fork out of loops

---------

Co-authored-by: Nico Flaig <nflaig@protonmail.com>
Co-authored-by: twoeths <tuyen@chainsafe.io>

nflaig committed:
fix: use attestation v1 endpoints pre-electra (#7024)

* fix: use attestation v1 endpoints pre-electra

* Adapt block attestations error message with pool apis

g11tech committed:
fix: use correct partial withdrawal index to evaluate processing them

twoeths committed:
fix: ssz v0.17.1 (#7035)

* fix: work around for sliceFrom() api

* Revert "fix: work around for sliceFrom() api"

This reverts commit 7aa6678cd46b1fa770f00923003a311a9d7a65da.

* fix: upgrade ssz

nflaig commented on packages/beacon-node/test/unit/util/sszBytes.test.ts:
```diff
@@ -1,43 +1,72 @@
-  getAggregationBitsFromAttestationSerialized,
+  getAttDataFromAttestationSerialized,
+  getAttDataFromSignedAggregateAndProofPhase0,
+  getAggregationBitsFromAttestationSerialized as getAggregationBitsFromAttestationSerialized,
```
```suggestion
  getAggregationBitsFromAttestationSerialized,
```

nflaig reviewed (dismissed):
LGTM - I don't see anything obvious that would break pre-electra

Do we wanna merge unstable into this branch instead of rebasing to resolve conflicts?

twoeths reviewed (commented):
`processEffectiveBalanceUpdates` has a performance issue, resolved it in #7043
need to merge that one and confirm that we don't have other regressions before merging this PR

twoeths commented:
e2e consistently failed in https://github.com/ChainSafe/lodestar/actions/runs/10518956587/job/29145618824?pr=7043
- test/e2e/chain/lightclient.test.ts > chain / lightclient > Lightclient track head on server configuration
- test/e2e/chain/proposerBoostReorg.test.ts > proposer boost reorg > should reorg a late block at slot 10
- test/e2e/sync/finalizedSync.test.ts > sync / finalized sync > should do a finalized sync from another BN

one error I see in `proposerBoostReorg.test.ts` is

```
Eph 1/5 0.028[vc-0-VAL-0-7]    error: Error publishing attestations slot=13 - submitPoolAttestations failed with status 500: Cannot read properties of undefined (reading 'submitPoolAttestationsV2')
Error: submitPoolAttestations failed with status 500: Cannot read properties of undefined (reading 'submitPoolAttestationsV2')
    at ApiResponse.apiResponse.error (/Users/tuyennguyen/Projects/lodestar/packages/beacon-node/test/utils/node/validator.ts:111:20)
    at ApiResponse.assertOk (/Users/tuyennguyen/Projects/lodestar/packages/api/src/utils/client/response.ts:155:18)
    at AttestationService.signAndPublishAttestations (/Users/tuyennguyen/Projects/lodestar/packages/validator/src/services/attestation.ts:243:78)
    at AttestationService.runAttestationTasksGrouped (/Users/tuyennguyen/Projects/lodestar/packages/validator/src/services/attestation.ts:155:5)
    at AttestationService.runAttestationTasks (/Users/tuyennguyen/Projects/lodestar/packages/validator/src/services/attestation.ts:117:9)
    at Clock.runAtMostEvery (/Users/tuyennguyen/Projects/lodestar/packages/validator/src/util/clock.ts:96:7)
```

cc @nflaig @ensi321

g11tech committed:
chore: rebase fixes

nflaig commented on packages/beacon-node/src/api/impl/validator/index.ts:
```diff
@@ -1076,8 +1078,16 @@ export function getValidatorApi(
 
-      const dataRootHex = toRootHex(attestationDataRoot);
-      const aggregate = chain.attestationPool.getAggregate(slot, dataRootHex);
+      const dataRootHex = toHex(attestationDataRoot);
```
```suggestion
      const dataRootHex = toRootHex(attestationDataRoot);
```

nflaig commented on packages/state-transition/src/block/processAttestationPhase0.ts:
we should probably rename this file to `processAttestation` as it handles phase0 + electra attestations now

nflaig committed:
chore: fix api handler used during testing (#7047)

twoeths commented:
I deployed #7043 to beta, metrics look good to me, all are comparable to unstable

twoeths committed:
fix: improve processEffectiveBalanceUpdates (#7043)

matthewkeil committed:
fix: perf test relative import from state-transition (#7055)

test: make import relative and add eslint-disable

g11tech committed:
fix: use next epoch for pending balance/consolidations processing (#7053)

wemeetagain commented:
:tada: This PR is included in [v1.22.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.22.0) :tada:
